XML compression

satyaakam opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->
https://dimagi-dev.atlassian.net/browse/IIO-742
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Changes in BlobDB API were made as per the suggestions on https://github.com/dimagi/commcare-hq/issues/26813
##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->
ICDS-CAS

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->
None

satyaakam committed:
XML compression

snopoke commented on corehq/blobs/s3db.py:
```diff
@@ -68,17 +69,25 @@ def put(self, content, **blob_meta_args):
         else:
             content.seek(0)
-            meta.content_length = get_file_size(content)
+            if meta.compressed:
+                meta.content_length = get_file_size(GzipCompressReadStream(content))
+                # Need to re-initialize stream as cannot seek on Gzipstream
+                content.seek(0)
+                content = GzipCompressReadStream(content)
```
can we avoid doing the compression twice?

```suggestion
            if meta.compressed:
                buf = io.BytesIO()
                with GzipFile(fileobj=buf, mode='wb') as f:
                    f.write(data)
                content = buf
            meta.content_length = get_file_size(content)
```
I realize that this will load all the data in memory but that's probably OK for our current use cases (at least it would simplify things initially and we could optimize if we need to compress larger objects). If we did this we could get rid of `GzipCompressReadStream` completely

millerdev replied:
@snopoke I would prefer to maintain support for arbitrarily large files. The blob db has been designed that way from the beginning, and I think it's an important part of the mental model for building things on top of it.

My recommendation for accomplishing that would be to have `GzipCompressReadStream` keep track of how many bytes have been read
```py
            if meta.compressed:
                content = GzipCompressReadStream(content)
            with self.report_timing('put', meta.key):
                s3_bucket.upload_fileobj(content, meta.key)
            meta.content_length = get_file_size(content)
            self.metadb.put(meta)
        return meta

...

def get_file_size(fileobj):
    if isinstance(fileobj, GzipCompressReadStream):
        return fileobj.content_length
```
`<GzipCompressReadStream>.content_length` implementation TBD. The attribute should not be set initially (it is an error to read the content length before all bytes have been read). It should be set to an integer value once all bytes have been read from the input.

Caveat: I can't remember if it's important to save the metadata _before_ writing blob content to S3. If it is then we have a big problem since we could never optimize this for arbitrary file size.

However, there is another question I'm not sure we've considered: is `meta.content_length` the number of bytes stored in the backend, or is it the number of bytes that would be returned by reading the file-like object returned by `blobdb.get(...)`? I can see reasons why we would want both, and it seems possibly important to have it be the latter since the blob db handles compression transparently. Thoughts?

snopoke replied:
> However, there is another question I'm not sure we've considered: is meta.content_length the number of bytes stored in the backend (compressed bytes), or is it the number of bytes that would be returned by reading the file-like object returned by blobdb.get(...) (uncompressed bytes)? I can see reasons why we would want both, and it seems possibly important and more intuitive to have it be the latter since the blob db handles compression transparently. Thoughts?

I was wondering about this too. We don't really use this field very much but in the few places where it is used we need the uncompressed size (except where we post metrics to Datadog where we'd want the compressed size). One place were we wouldn't be able to get the uncompressed size is where we copy an existing blob: https://github.com/dimagi/commcare-hq/blob/8f63335b220751b0bdaac9e13abb818e0ab02413/corehq/blobs/s3db.py#L70. This appears to have been added specifically for case attachments here: https://github.com/dimagi/commcare-hq/pull/13203. I couldn't find any usages of it but there must be some since we're recording metrics for it on [production](https://app.datadoghq.com/dashboard/753-35q-fwt/blob-db-dashboard?from_ts=1584428385517&fullscreen_section=overview&fullscreen_widget=3073215485582670&live=true&tile_size=m&to_ts=1587020385517&tpl_var_env=production&utc_override=true&fullscreen_start_ts=1584446132694&fullscreen_end_ts=1587038132694&fullscreen_paused=false). Either way that does seem like a useful feature to keep. One approach would be to force the user to pass in the `meta` to `db.put` if they are passing in an existing blob or else pass the content length info into `BlobStream`.

I also noticed that if we return a `GzipFile` from `db.get` then the type check `isinstance(content, BlobStream)` will fail and we'll end up to a read then write.

snopoke replied:
@millerdev I've made some updates here so that `meta.content_length` is the uncompressed size but there are 2 remaining challenges:

1. `db.size(key) != meta.content_length`

This is used in a few places to to add content length headers to responses. I think this could be solved by applying the same arg rules to `size` as we have to `get`. We could also eliminate all calls to `size` if we return the content length along with the blob from `get` since they are all in places where the actual blob is fetched as well.

2. The metrics are reporting the uncompressed size which I think is incorrect.

Not clear how to solve this since many places would not be able to get the compressed size easily e.g. bulk delete

millerdev replied:
Have you considered adding a `compressed_size` attribute to `BlobMeta`? It would be null if `<BlobMeta>.compressed` is false. We could even improve storage efficiency a bit by removing `compressed` entirely and make the rule be that if `compressed_size` is null then the blob is not compressed, otherwise it is compressed. Thoughts?

Storing both `compressed_size` in addition to `content_length` is beginning to seem more straightforward than other options.

snopoke commented on corehq/blobs/interface.py:
```diff
@@ -52,7 +53,7 @@ def put(self, content, **blob_meta_args):
     @abstractmethod
-    def get(self, key):
+    def get(self, key=None, type_code=None, meta=None):
         """Get a blob
```
should update the docstring with details of how to call this

snopoke commented on corehq/blobs/interface.py:
```diff
@@ -61,6 +62,25 @@ def get(self, key):
+            return key
+        if meta is None:
+            raise ValueError("'key' and 'type_code' or 'meta' is required")
+        return meta.key
```
I think the version here is more concise: https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727

satyaakam replied:
Following issues with that version:
- If function is called with `type_code == "form_xml"` and one of `key` or `error`, they will get following error:
```
'key' must be specified with 'type_code'
```
whereas what they really should get is
```
form XML must be loaded with 'meta' argument
```
Current version, although longer, handles all possible cases with the correct error message

millerdev replied:
I didn't follow this. Are you saying this function call
```py
validate_get_args("some-key", CODES.form_xml, None)
```
will raise `'key' must be specified with 'type_code'`?

That's not what I see. Tested locally this is what I get:
```
Traceback (most recent call last):
  File "<string>", in validate_get_args
ValueError: form XML must be loaded with 'meta' argument

```

satyaakam replied:
No, I am referring to the version at https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727 since @snopoke suggested using it.

Using that version will raise `'key' must be specified with 'type_code'` with following function call:

validate_get_args("some-key", CODES.form_xml, None)

millerdev replied:
Might want to check that again. I ran the version from https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727 and got `ValueError: form XML must be loaded with 'meta' argument`

```py
>>> class CODES:
...     # stub
...     form_xml = 2
...
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
...
>>> validate_get_args("some-key", CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 8, in validate_get_args
ValueError: form XML must be loaded with 'meta' argument
```

satyaakam replied:
Ok, I checked again. But consider below

>>> class CODES:
...     form_xml = 2
... 
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
... 
>>> validate_get_args(None, CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in validate_get_args
ValueError: 'key' must be specified with 'type_code'

satyaakam replied:
Ok, I checked again. But consider below

```
>>> class CODES:
...     form_xml = 2
... 
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
... 
>>> validate_get_args(None, CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in validate_get_args
ValueError: 'key' must be specified with 'type_code'
```

millerdev replied:
Looks right to me (you sent a null key with a type code). Do you see a problem?

satyaakam replied:
"We sent `type_code` as `form_xml`. `key` is not expected for this type_code and only `meta` should be provided. Yet the error asks me to provide `key`."

millerdev replied:
This should fix it
```py
def validate_get_args(key, type_code, meta):
    if key is not None or type_code is not None:
        if meta is not None:
            raise ValueError("'key' and 'meta' are mutually exclusive")
        if type_code == CODES.form_xml:
            raise ValueError("form XML must be loaded with 'meta' argument")
        if key is None or type_code is None:
            raise ValueError("'key' must be specified with 'type_code'")
        return key
    if meta is None:
        raise ValueError("'key' and 'type_code' or 'meta' is required")
    return meta.key
```

snopoke commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,57 @@ def pre_init(self, value, obj):
+        pass
+
+
+class GzipCompressReadStream(object):
```
This could do with some tests. Also wondering if it wouldn't be cleaner to implement this as a file wrapper using the zlip encoder directly. Something similar to this: http://effbot.org/librarybook/zlib/zlib-example-4.py

satyaakam replied:
test cases written in test_util.py

snopoke commented on corehq/blobs/s3db.py:
```diff
@@ -68,17 +69,25 @@ def put(self, content, **blob_meta_args):
                 s3_bucket.upload_fileobj(content, meta.key)
         return meta
 
     def get(self, key):
```
need to add the args here

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -1070,6 +1074,7 @@ class CaseAttachmentSQL(PartitionedModel, models.Model, SaveStateMixin, IsImageM
 
     # DEPRECATED - use CaseAttachmentSQL.content_md5() instead
     md5 = models.CharField(max_length=255, default="")
+    type_code = models.IntegerField(default=CODES.form_attachment)
```
I don't think this is necessary as `CaseAttachments` will always be `CODES.form_attachment`. Also we don't really support case attachments (it's on the kill list)

snopoke reviewed (commented):
I did a quick pass and left some comments. Overall looks like a good start.

I don't think you meant to commit these:
![image](https://user-images.githubusercontent.com/249606/78132559-3927ba80-741d-11ea-945f-09c77fd42046.png)

millerdev commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,57 @@ def pre_init(self, value, obj):
+    def read(self, size=-1):
+        if size < 0: size = self.__size
+        ret_list = []
+        while size > 0 and len(self.__buf):
```
```suggestion
        while size > 0 and self.__buf:
```

millerdev commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,57 @@ def pre_init(self, value, obj):
+class Buffer(object):
+    def __init__(self):
+        self.__buf = deque()
+        self.__size = 0
```
Private members should be prefixed with one underscore, not two. Name mangling (invoked by dunder-prefix) is nasty and should be avoided if possible. The same applies to `GzipCompressReadStream` below.

millerdev commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,57 @@ def pre_init(self, value, obj):
 
 
+# adapted from https://stackoverflow.com/a/31566082
+class Buffer(object):
```
nit: Python 3 ftw!
```suggestion
class Buffer:
```

millerdev commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,57 @@ def pre_init(self, value, obj):
+        pass
+
+    def close(self):
+        pass
```
```suggestion
        self.__buf = None
```

millerdev commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,57 @@ def pre_init(self, value, obj):
+        self.__gzip = GzipFile(None, mode='wb', fileobj=self.__buf)
+
+    def read(self, size=-1):
+        while size < 0 or len(self.__buf) < size:
```
```suggestion
        while size < 0 or self.__buf < size:
```

millerdev replied:
Ignore this. I don't know what I was thinking here.

satyaakam committed:
Changes made as per the suggestions in PR 27022

satyaakam committed:
Resolving conflits

satyaakam committed:
merged master into compress-xml

satyaakam committed:
Merge branch 'master' of https://github.com/dimagi/commcare-hq into compress-xml

satyaakam committed:
Changes to the compress xml as review

satyaakam committed:
Merge branch 'master' of https://github.com/dimagi/commcare-hq into compress-xml

snopoke commented on corehq/blobs/migrations/0011_blobmeta_compressed.py:
```diff
@@ -0,0 +1,20 @@
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
must be annotated with `@partitioned` (`corehq.sql_db.migrations.partitioned`)

satyaakam committed:
Fixing fucntion call

millerdev commented on corehq/blobs/s3db.py:
```diff
@@ -189,6 +196,10 @@ def is_not_found(err, not_found_codes=["NoSuchKey", "NoSuchBucket", "404"]):
 
 
 def get_file_size(fileobj):
+    CHUNK_SIZE = 4096
```
Is this used?

millerdev commented on corehq/blobs/util.py:
```diff
@@ -37,6 +39,79 @@ def pre_init(self, value, obj):
+        @property
+        def content_length(self):
+            if self._content_length is None or self._size > 0:
+                raise Exception("content_length can't be accessed without completely reading the stream")
```
This should use a custom exception subclass rather than `Exception`. As is, it is impossible to catch only this error.

millerdev commented on corehq/blobs/tests/test_util.py:
```diff
@@ -16,3 +19,32 @@ def test_random_id_length(self):
+                compress_stream.content_length
+            content_length = len(compress_stream.read())
+            self.assertEqual(compress_stream.content_length, content_length)
+
```
Please add another test: write several bytes, then do a partial read (say write 10 bytes, then read 5), verify that exception is raised on access `compress_stream.content_length` before all 10 bytes have been read.

satyaakam committed:
Missing steps to make Bower work as normal user

satyaakam committed:
Merge branch 'master' of https://github.com/dimagi/commcare-hq

satyaakam committed:
Updated Tests and custom exceptions added

satyaakam committed:
Merge branch 'master' into compress-xml

satyaakam committed:
Fixed Lint issues

snopoke committed:
simplify validate_args

snopoke committed:
nits

snopoke committed:
add missing type_code kwargs

snopoke committed:
remove unused imports

snopoke committed:
handle exception

snopoke committed:
more type_code args + test fixes

snopoke committed:
make meta.content_length = uncompressed size

snopoke committed:
uncompressed content_length when copying a blob

snopoke committed:
test with compression

snopoke committed:
Merge branch 'master' into compress-xml

millerdev commented on corehq/blobs/models.py:
```diff
@@ -93,13 +93,18 @@ def is_image(self):
         """
-        from . import get_blob_db
-        return get_blob_db().get(meta=self)
+        from . import get_blob_db, CODES
+        if self.type_code == CODES.form_xml:
+            kwargs = {'meta': self}
+        else:
+            kwargs = {'key': self.key, 'type_code': self.type_code}
+        db = db or get_blob_db()
+        return db.get(**kwargs)
```
This can be simplified
```suggestion
        from . import get_blob_db
        db = db or get_blob_db()
        return db.get(meta=self)
```

millerdev reviewed (approved):
Looking good. One small recommendation.

snopoke committed:
store compressed length and always return a BlobStream instance

snopoke committed:
simplify `blobmeta.open()`

Co-Authored-By: Daniel Miller <millerdev@gmail.com>

snopoke committed:
reusue util to get sizes

snopoke committed:
minor simplication

snopoke committed:
prefer using meta for lookup if it's available

snopoke committed:
better test

snopoke commented on corehq/blobs/metadata.py:
```diff
@@ -31,6 +32,8 @@ def new(self, **blob_meta_args):
             return blob_meta_args["meta"]
         timeout = blob_meta_args.pop("timeout", None)
+        if blob_meta_args.get('type_code') == CODES.form_xml:
+            blob_meta_args['compressed_length'] = -1
```
I don't really like this. It does got overwritten during later but still feels off

snopoke commented on corehq/blobs/metadata.py:
```diff
@@ -17,6 +17,7 @@ class MetaDB(object):
 
     This class manages persistence of blob metadata in a SQL database.
     """
+    DoesNotExist = BlobMeta.DoesNotExist
```
Used here: https://github.com/dimagi/commcare-hq/blob/e0d7c553ea27a08146acbb9b4b4a9a39753b7b16/corehq/blobs/mixin.py#L193

snopoke commented on corehq/blobs/s3db.py:
```diff
@@ -65,25 +63,42 @@ def put(self, content, **blob_meta_args):
             with self.report_timing('put', meta.key):
-                s3_bucket.upload_fileobj(content, meta.key)
+                s3_bucket.upload_fileobj(content, meta.key, Callback=_track_transfer)
+            meta.content_length, meta.compressed_length = get_content_size(content, chunk_sizes)
```
I don't thinks so:
```python

def get_content_size(fileobj, chunks_sent):
    if isinstance(fileobj, GzipCompressReadStream):
        return fileobj.content_length, sum(chunks_sent)

    return sum(chunks_sent), None
```
So when it's compressed then it returns the uncompressed size and the sum of the chunks sent which should be the compressed size.

snopoke replied:
Note that the old implementation doesn't work in most cases since the file get's closed by `s3_bucket.upload_fileobj`.

millerdev replied:
`fileobj.content_length` (where `fileobj` is `GzipCompressReadStream`) returns compressed size

https://github.com/dimagi/commcare-hq/blob/820f81d1249157f34b592bebdb5d6b8868ae461f/corehq/blobs/util.py#L58-L60

snopoke replied:
That is `file_obj._buf.content_length` which can actually be removed now: 6d2f0c00c9c

snopoke committed:
Update corehq/blobs/interface.py

Co-Authored-By: Daniel Miller <millerdev@gmail.com>

snopoke committed:
fix argument order

snopoke commented:
That is `file_obj._buf.content_length` which can actually be removed now.

On Fri, 17 Apr 2020, 17:20 Daniel Miller, <notifications@github.com> wrote:

> *@millerdev* commented on this pull request.
> ------------------------------
>
> In corehq/blobs/s3db.py
> <https://github.com/dimagi/commcare-hq/pull/27022#discussion_r410293646>:
>
> >              with self.report_timing('put', meta.key):
> -                s3_bucket.upload_fileobj(content, meta.key)
> +                s3_bucket.upload_fileobj(content, meta.key, Callback=_track_transfer)
> +            meta.content_length, meta.compressed_length = get_content_size(content, chunk_sizes)
>
> fileobj.content_length (where fileobj is GzipCompressReadStream) returns
> compressed size
>
>
> https://github.com/dimagi/commcare-hq/blob/820f81d1249157f34b592bebdb5d6b8868ae461f/corehq/blobs/util.py#L58-L60
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/27022#discussion_r410293646>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB46BVFQMYBBKVK7VUKVKDRNBXU5ANCNFSM4LYXQXGQ>
> .
>

satyaakam committed:
merging upstream changes

satyaakam committed:
Merging upstream changes

snopoke committed:
remove GzipCompressReadStream.Buffer.content_length

snopoke committed:
Merge branch 'master' into compress-xml

millerdev committed:
GzipCompressReadStream -> GzipStream

millerdev committed:
GzipStreamAttrAccessBeforeRead -> GzipStreamError

Make the error more generic so it can be used in other places if need be. Currently this error is never expected to occur unless there is a bug.

millerdev committed:
Move IO buffer out of GzipStream class

Both GzipStream and GzipStream.Buffer had a `_buf` attribute. The class composition made it easy to confuse the two.

millerdev committed:
Fix bugs in GzipStream

millerdev reviewed (commented):
I fixed some bugs and created a new PR against the original branch in @satyaakam's fork: https://github.com/satyaakam/commcare-hq/pull/1

@snopoke I'm not sure how you were able to push directly to this PR since it's originating from a fork. I'd be interested in learning how you did that.

snopoke committed:
Merge pull request #1 from dimagi/dm/compress-xml

Gzip stream fixes for XML compression

snopoke commented:
> I'm not sure how you were able to push directly to this PR since it's originating from a fork. I'd be interested in learning how you did that.

I got given write access to the forked repo

snopoke committed:
save to file before extracting in test

snopoke committed:
fallback to partial read if seek not supported

snopoke commented:
@millerdev build is passing but will let you do a final review before merging

shubham1g5 commented:
Flagging that this PR incorrectly updated the commcare-translations submodule reverting it back to an older 2.47 version. @orangejenny has made the required correction on master pushing it back to the latest commit.
